src/npc_attack.cpp

```diff
@@ -298,17 +359,26 @@ npc_attack_rating npc_attack_gun::evaluate_tripoint(
     }
     const int distance_to_me = rl_dist( location, source.pos() );
     const Creature::Attitude att = source.attitude_to( *critter );
-    const bool friendly_fire = att == Creature::Attitude::FRIENDLY &&
-                               !source.rules.has_flag( ally_rule::avoid_friendly_fire );
-    int attitude_mult = npc_attack_constants::attitude_multiplier.at( att );
-    if( friendly_fire ) {
-        // hitting a neutral creature isn't exactly desired, but it's a lot less than a friendly.
-        // if friendly fire is on, we don't care too much, though if an available hit doesn't damage them it would be better.
-        attitude_mult = npc_attack_constants::attitude_multiplier.at( Creature::Attitude::NEUTRAL );
-    }
-    const int distance_penalty = std::max( distance_to_me - 1,
-                                           ( source.closest_enemy_to_friendly_distance() - 1 ) * 2 );
-    double potential = damage * attitude_mult - distance_penalty;
+
+    if( att != Creature::Attitude::HOSTILE ) {
+        // No point in throwing stuff at neutral/friendly targets
+        return npc_attack_rating( cata::nullopt, location );
+    }
+
+    bool avoids_friendly_fire = source.rules.has_flag( ally_rule::avoid_friendly_fire );
+
+    double potential = damage;
+
+    // Make attacks that involve moving to find clear LOS slightly less likely
+    if( has_obstruction( source.pos(), location, avoids_friendly_fire ) ) {
+        potential *= 0.9f;
+    } else if( avoids_friendly_fire && !source.wont_hit_friend( location, gun, false ) ) {
+        potential *= 0.95f;
+    }
+
+    // Prefer targets not directly next to you
```

Comment:
why? npcs still have to deal with those

Reply:
Trying to reduce weights for closer targets so melee weapons can be a preferred choice for those targets even if the damage would be lower. Shooting point blank comes with issues as you can't aim properly.